### PR TITLE
WT-9455 Refactor Cmake and Evergreen defaults

### PR DESCRIPTION
**Pull request by mickgraham-mongodb**
Updating cmake and evergreen defaults to have less duplication.

Making the default more developer friendly and more likely to match evergreen.

### CODE REVIEW

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:762]: I like putting the configure flags in variables, but I'm not sure about the "variables that include other variables" thing. Stuff like builtins and static are orthogonal, and I think it'd be clearer to have tasks use different variables for different bits of config. Stuff like `configure_flags_address_sanitizer_mongodb_v4_clang` below are IMO a bit much (the name also doesn't include the fact that it uses builtins, which would make it even more difficult to read).

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:1144]: I can see that a couple of cppsuite tasks got removed, is that a temporary thing? I can't see a replacement anywhere but I might be missing it in the monster diff for this file.

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:178]: Shouldn't we just be using `- func: "fetch artifacts"` here? I don't think there's a need to give the function a variable name, it's only used in one spot.

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:2622]: Factoring out the common bits is a nice improvement here, thanks.

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:170]: I didn't think `-DWITH_PIC=1` was a default, so we can't remove these (and the ones below)? I might be missing something here.

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:4096]: I know you weren't convinced about the TCMalloc stuff, but it looks OK? Sorry that the issue (if any) isn't jumping out at me.

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:4402]: What was this doing? It's not obvious to me that we can remove it.

**Review by korteland — COMMENTED** [review]
This is heading in a really good direction, thanks for paying some attention to this. (One final nitpick, the PR title should be "WT-9455 ..." rather than "Wt 9455...".)

File: CMakeLists.txt
Comment (mickgraham-mongodb) [reply CMakeLists.txt:64]: For windows and the version of cmake we are using, it definitely did the wrong thing and added -std:c++17 which doesn't pick up what we need (needs to be /std:c++17 from what I understand for MSVC). 

Otherwise I wanted for the other platforms/compilers to make it as close to MongoDB as possible. Apart from the tiered storage extension which is part of the shipping code, only test code uses C+ so we could argue that it doesn't really need to match but I think it is good to do it. This is why I have the extensions turned off.

File: CMakeLists.txt
Comment (mickgraham-mongodb) [reply CMakeLists.txt:71]: See previous comment reply. Some of it is about aligning with MongoDB and the other part is compatibility with windows.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:18]: This is was more about alignment with what evergreen was doing. Sanitizer builds had HAVE_DIAGNOSTICS=1, but not a lot of the others. I think we could decide for our default to always HAVE_DIAGNOSTICS=1. Note that debug (e.g. -ggdb) is on all builds by default except Release. 

My understanding with HAVE_DIAGNOSTICS is it changed behaviour more than just debug symbols.

File: cmake/configs/modes.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/modes.cmake:223]: If we don't remove them then we don't control the optimisation level via our flag which is what we wanted. This change was to make it consistent across all the build types.

File: test/unittest/tests/test_reconciliation_tracking.cpp
Comment (mickgraham-mongodb) [reply test/unittest/tests/test_reconciliation_tracking.cpp:31]: This was to fix a memory leak that was now found because we enabled unittests with the sanitizer builds via the PR. If this should be different, we can pull that out and fix it first. Jeremy made this fix.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:762]: I can update the name to make it clearer. But with evergreen, you can't have two uses of vars (<<: syntax) in one section so you can't use variables cleanly in a way. Its actually possible for us to think about making the sanitizer a build variant and make that much cleaner (by putting the build type in the variant definition) which would stop some of this duplication. I can look at that.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:170]: It should have been a default from how we were actually using it from what I could tell. That was one of the changes I made in base.cmake.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:4096]: In the patch build I'm not convinced we pick up that we have TCMALLOC_LIB. That is something I need to confirm.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:4402]: It was redundant and in the windows powershell script directly.

Comment (korteland) [reply test/evergreen.yml:4402]: Cool, thanks.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml:4096]: OK, if you want a hand debugging then shout out.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml:170]: Ah you're right, I missed that small off -> on change. Thanks for clarifying.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml:762]: Thanks - that sounds like a better way to do it. That's a shame about not having multiple variables though.

File: cmake/configs/modes.cmake
Comment (korteland) [reply cmake/configs/modes.cmake:223]: Cool, thanks for explaining that.

File: cmake/configs/base.cmake
Comment (korteland) [reply cmake/configs/base.cmake:18]: Yeah, `HAVE_DIAGNOSTICS` includes a tonne of extra checks, some of them change behaviour/performance quite a bit. Thinking about it a bit more, I kinda think Evergreen was doing the wrong thing before - they probably should've been on. Do you know if there was a reason that only the sanitisers enabled diagnostics?

File: CMakeLists.txt
Comment (korteland) [reply CMakeLists.txt:64]: OK, that's interesting. Do you know which versions of CMake + MSVC you used? I read some documentation, and my understanding is that you need cmake >= 3.10 and MSVC >= 2017 for this to work. Maybe we need to clarify what the oldest supported MSVC is.

Regarding the C++ extensions, I didn't realise that MongoDB doesn't use them. If we're currently fine without extensions then yeah I agree it makes sense to stick to the standard.

File: test/unittest/tests/test_reconciliation_tracking.cpp
Comment (korteland) [reply test/unittest/tests/test_reconciliation_tracking.cpp:31]: OK, I'm now tracking this is WT-9548.

File: cmake/configs/base.cmake
Comment (etienneptl) [reply cmake/configs/base.cmake:276]: I believe we are referring to WT-9481: `FIXME-WT-9481`. It should be captured by [s_mentions](https://github.com/wiredtiger/wiredtiger/blob/develop/dist/s_mentions) when someone works on that ticket.

**Comment by mickgraham-mongodb** [comment]
Not yet ready for further review. Will update based on Will's and Etienne's comments first, fix the TCMalloc issue, and then notify when its ready.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:282]: Done.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:48]: I'll add a new variant that is sans all the extensions and python tests. Note these aren't enabled on Windows.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:275]: Done.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:93]: Done.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:364]: Done.

File: cmake/configs/modes.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/modes.cmake:191]: Fixed.

File: cmake/configs/modes.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/modes.cmake:207]: Fixed

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:1144]: It was a mistake. Great catch.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:181]: Can't do because its in the functions section I believe.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:2625]: Thanks!

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml]: Found it, was based on the directory when the var was resolved. Fixed!

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:18]: As discussed, changing to have diagnostics on by default except for release builds. Also turning off for perf builds in evergreen.

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:765]: There isn't a clean path because the tests that use the sanitisers aren't always in a sanitiser variant. Updated the naming to at least be consistent.

**Comment by mickgraham-mongodb** [comment]
Ready for review now.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml]: Nice!

File: cmake/configs/base.cmake
Comment (korteland) [reply cmake/configs/base.cmake:18]: Thanks, I think it's a bit easier to read now.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml]: Re-reviewing, I don't think this is fixed yet? At least,  I still can't find `burst_inserts_default` in this file.

File: cmake/configs/base.cmake
Comment (korteland) [reply cmake/configs/base.cmake:48]: Thanks, the new `ubuntu2004-minimal` variant looks good.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml:765]: Thanks. It's better now, shame about the lack of multiple variables but nothing we can do about it in this PR.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml:183]: Ack, thanks for looking into it.

File: test/evergreen.yml
Comment (korteland) [reply test/evergreen.yml]: Nevermind, I typo'ed. Thanks for fixing it!

**Review by korteland — APPROVED** [review]
This looks good to me now, thanks for improving this! Looks like the new minimal variant is running the python tests, but I think that's the only issue remaining.

**Comment by mickgraham-mongodb** [comment]
There were two fixes that were required (missing cppsuite test that Will caught correctly and the python test in the minimal variant). Fixed those and now testing again.

File: cmake/configs/modes.cmake
Comment (etienneptl) [reply cmake/configs/modes.cmake:223]: Still missing a new line @mickgraham-mongodb

File: ext/storage_sources/s3_store/s3_log_system.h
Comment (etienneptl) [inline ext/storage_sources/s3_store/s3_log_system.h:85]: Should we add the `final` keyword as well?
```suggestion
    Flush() override final
```

**Review by agorrod — APPROVED** [review]
This is a big change, but from my review it looks like a big improvement. Thanks!
I'm going to approve, assuming that you will resolve any known outstanding issues - let me know if I can help with them.

**Review by etienneptl — COMMENTED** [review]
LGTM!
When I run the `validate` command, I have the following warning that should be considered:
```
evergreen validate test/evergreen.yml 
WARNING: task 'unit-test-long' in buildvariant 'ubuntu2004-arm64' references a nonexistent distro or container named 'ubuntu2004-arm64-medium'
test/evergreen.yml is valid with warnings
```

File: test/evergreen.yml
Comment (mickgraham-mongodb) [reply test/evergreen.yml:197]: Its just a diff issue. It was always silent (for the tiered storage tests). Yes its about the creds.

File: cmake/README.md
Comment (mickgraham-mongodb) [reply cmake/README.md:99]: I think encryptor is consistent with compressor.

File: ext/storage_sources/s3_store/s3_log_system.h
Comment (mickgraham-mongodb) [reply ext/storage_sources/s3_store/s3_log_system.h:85]: I thought it was either override or final. Do you mean const override?

Comment (etienneptl) [reply ext/storage_sources/s3_store/s3_log_system.h:85]: It is not blocking the PR but it's something I noticed. If we don't intend the inherit from that class and override `Flush`, we could have marked it as `final` too.

**Review by etienneptl — APPROVED** [review]
LGTM! Awesome changes!

File: test/evergreen.yml
Comment (etienneptl) [inline test/evergreen.yml:4208]: `ubuntu2004-medium` should be fine.

File: ext/storage_sources/s3_store/s3_log_system.h
Comment (mickgraham-mongodb) [reply ext/storage_sources/s3_store/s3_log_system.h]: Done.

**Pull request merged**
